{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3437fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib       \n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np      \n",
    "import pandas as pd     \n",
    "import rushd as rd      \n",
    "import scipy as sp      \n",
    "import seaborn as sns   \n",
    "\n",
    "# Set seaborn style\n",
    "sns.set_style('ticks')\n",
    "sns.set_context('talk', rc={'font.family': 'sans-serif', 'font.sans-serif':['Helvetica Neue']})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "730b4fe0",
   "metadata": {},
   "source": [
    "There are two examples in this notebook.\n",
    "\n",
    "1. Measuring relative expression of transcripts using RT-qPCR, example data from ComMAND\n",
    "2. Quantifying DNA library prep for a genomics assay (RCMC), example data from TANGLES"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99087949",
   "metadata": {},
   "source": [
    "# Example 1: Measuring relative expression across conditions for RT-qPCR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "941dfb10",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b43ba000",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = rd.datadir/'data'/'qPCR'/'emma'/'command'\n",
    "output_path = rd.rootdir/'output'/'qPCR-example'\n",
    "cache_path = output_path/'data.gzip'\n",
    "\n",
    "plates = pd.DataFrame({\n",
    "    'data_path': [base_path/'2024.11.13_command'/'2024.11.13_ELP_command_qPCR_Cp.txt', base_path/'2024.11.22_command'/'2024.11.22_ELP_command_qPCR_Cp.txt', base_path/'2024.12.03_command'/'2024.12.03_ELP_command_qPCR_Cp.txt'],\n",
    "    'yaml_path': [base_path/'2024.11.13_command'/'wells.yaml', base_path/'2024.11.22_command'/'wells.yaml', base_path/'2024.12.03_command'/'wells.yaml'],\n",
    "    'exp': ['exp123', 'exp123.2', 'exp123.4']\n",
    "})\n",
    "\n",
    "for p in plates['yaml_path'].unique():\n",
    "    rd.plot.plot_well_metadata(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0770630c",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.DataFrame()\n",
    "if cache_path.exists(): data = pd.read_parquet(cache_path)\n",
    "else: \n",
    "    data = rd.qpcr.load_plates_with_metadata(plates, is_default=True)\n",
    "    data.to_parquet(rd.outfile(cache_path))\n",
    "data.dropna(subset=['primers'], inplace=True)\n",
    "display(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bbc9b5d",
   "metadata": {},
   "source": [
    "## Use controls to exclude low-quality measurements, combine technical reps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a59e6f85",
   "metadata": {},
   "source": [
    "Look at controls first.\n",
    "\n",
    "We have negative controls without any cDNA (no template for reaction) and without reverse transcriptase (for the RT reaction, so also should have no template). These set the level of \"background\" cycle counts, so we should treat conditions with Cp values greater than or equal to this threshold as not expressing.\n",
    "\n",
    "As another control, we have uninfected cells (\"UI\"), which should have detectable levels of endogenous transcripts (GAPDH, FMRP (real name: Fmr1), and FXN) but undetectable transgenes (mRuby2, EGFP). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0cf9f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = data[data['construct'].isin(['UI','no-RT', 'no-cDNA'])]\n",
    "g = sns.catplot(data=plot_df, x='primers', y='Cp', col='biorep', hue='construct', dodge=True)\n",
    "\n",
    "for ax in g.axes.flatten():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "271f2557",
   "metadata": {},
   "source": [
    "Now look at conditions of interest.\n",
    "\n",
    "Want to look at all technical replicates, in case there are outliers that we should exclude (i.e., wells with no amplification, Cp = 35)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57fc23f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = data[~data['construct'].isin(['UI','no-RT', 'no-cDNA'])]\n",
    "g = sns.catplot(data=plot_df, x='primers', y='Cp', row='biorep', col='construct', hue='dox',\n",
    "                dodge=True, margin_titles=True)\n",
    "for ax in g.axes[-1,:].flatten():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126a9fdc",
   "metadata": {},
   "source": [
    "Indeed, there are a few wells with no amplification. Let's remove all these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da59eaf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filtered = data[(data['Cp']<35)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5be758bd",
   "metadata": {},
   "source": [
    "Now, combine technical replicates. Usually, we take the median."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cb075d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats = data_filtered.groupby(['exp','biorep','construct','dox','primers'])[['Cp']].median().reset_index().dropna(subset='Cp')\n",
    "display(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f20dc8b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = stats[~stats['construct'].isin(['no-RT', 'no-cDNA'])]\n",
    "g = sns.catplot(data=plot_df, x='primers', y='Cp', col='construct', hue='dox',\n",
    "                dodge=True, margin_titles=True)\n",
    "for ax in g.axes.flatten():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c721aa2",
   "metadata": {},
   "source": [
    "For each set of primers, find conditions (construct + dox) where Cp values are above the no cDNA and no RT controls for a given biorep. This represents background amplification, so exclude these conditions.\n",
    "\n",
    "Not all reps and primer sets included these negative controls. In this case, we won't exclude anything. (We achieve this easily by setting the negative control Cps to 35, assuming no amplification and thus including all other Cp values)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3709d4da",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_on = ['exp','biorep','primers','construct']\n",
    "no_cdna_control = stats.set_index(index_on)[['Cp']].xs('no-cDNA', level='construct').rename(columns={'Cp': 'Cp_no-cDNA'})\n",
    "no_rt_control = stats.set_index(index_on)[['Cp']].xs('no-RT', level='construct').rename(columns={'Cp': 'Cp_no-RT'})\n",
    "stats = stats.merge(no_cdna_control, on=index_on[:-1], how='left').merge(no_rt_control, on=index_on[:-1], how='left')\n",
    "\n",
    "# not all reps + primers had these negative controls, so fill NaNs with Cp=35\n",
    "stats.fillna(35, inplace=True)\n",
    "\n",
    "# exclude conditions with high Cps as well as the control conditions\n",
    "stats_filtered = stats[(stats['Cp'] < stats['Cp_no-cDNA']) & ((stats['Cp'] < stats['Cp_no-RT'])) & ~(stats.construct.isin(['no-cDNA','no-RT']))].copy()\n",
    "display(stats_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f26b2d10",
   "metadata": {},
   "source": [
    "Calculate GAPDH expression for each condition (biorep + construct + dox).\n",
    "\n",
    "Then one more filtering step: Exclude conditions with low GAPDH expression. This indicates poor RNA extraction (or otherwise low cDNA template)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8840aa74",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_on = ['exp','biorep','construct','dox','primers']\n",
    "gapdh = stats_filtered.set_index(index_on)[['Cp']].xs('GAPDH', level='primers').rename(columns={'Cp': 'Cp_GAPDH'})\n",
    "stats_filtered = stats_filtered.merge(gapdh, on=index_on[:-1], how='left')\n",
    "stats_filtered.fillna(35, inplace=True)\n",
    "\n",
    "gapdh_thresh = 21   # set by eye / what is expected for a highly expressed gene\n",
    "stats_filtered = stats_filtered[stats_filtered['Cp_GAPDH'] < gapdh_thresh]\n",
    "display(stats_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e2bad75",
   "metadata": {},
   "source": [
    "Finally we have values for each primer set for each condition (biorep + construct + dox)! Let's plot to double-check."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c93e199",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = stats_filtered\n",
    "g = sns.catplot(data=plot_df, x='primers', y='Cp', col='construct', hue='dox',\n",
    "                dodge=True, margin_titles=True)\n",
    "for ax in g.axes.flatten():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42b85f95",
   "metadata": {},
   "source": [
    "## Calculate normalized expression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d1e0880",
   "metadata": {},
   "source": [
    "First, calculate $\\Delta C_p$ relative to the GAPDH condition. This normalizes the RNA extraction / RT reaction and differences in global expression across conditions.\n",
    "\n",
    "The normalized expression value is $2^{- \\Delta C_p}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0828e613",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_filtered['delta_Cp'] = stats_filtered['Cp'] - stats_filtered['Cp_GAPDH']\n",
    "stats_filtered['expression_norm'] = 2**(-stats_filtered['delta_Cp'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f2491a0",
   "metadata": {},
   "source": [
    "What we really care about is how the different conditions compare to each other. This is $\\Delta \\Delta C_p$, or relative expression.\n",
    "\n",
    "For ComMAND, we'll compare normalized expression for the circuit relative to the base gene (for each biorep + dox). We tested several circuits here, so we'll add metadata to annotate circuit type based on construct name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de2ef5fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(rd.rootdir/'inputs'/'plasmid-metadata.csv')\n",
    "display(metadata)\n",
    "\n",
    "stats_filtered = stats_filtered.merge(metadata, how='left', on='construct')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8039fc3d",
   "metadata": {},
   "source": [
    "Conveniently, we used different primer sets for each circuit. So for each primer set, there is only one base gene in the constructs tested.\n",
    "\n",
    "However, there are a few cases where the base gene is not included (e.g., we excluded that condition earlier because of high Cp values). In this case, let's exclude circuits missing measurements for the corresponding base gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f019dad",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_on = ['exp','biorep','dox','primers','group']\n",
    "base_delta_cp = stats_filtered.set_index(index_on)[['delta_Cp']].xs('base', level='group').rename(columns={'delta_Cp': 'delta_Cp_base'})\n",
    "stats_filtered = stats_filtered.merge(base_delta_cp, on=index_on[:-1], how='left')\n",
    "\n",
    "stats_circuits = stats_filtered.dropna(subset='delta_Cp_base')\n",
    "display(stats_circuits)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6736eded",
   "metadata": {},
   "source": [
    "Finally, we can compute $\\Delta \\Delta C_p$ and relative expression, $2^{-\\Delta \\Delta C_p}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcac92af",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_circuits['delta_delta_Cp'] = stats_circuits['delta_Cp'] - stats_circuits['delta_Cp_base']\n",
    "stats_circuits['relative_expression'] = 2**(-stats_circuits['delta_delta_Cp'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a3d3d18",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = stats_circuits\n",
    "g = sns.catplot(data=plot_df, x='construct', y='delta_delta_Cp', col='primers', hue='dox',\n",
    "                dodge=True, margin_titles=True)\n",
    "for ax in g.axes.flatten():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "464830d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_circuits['circuit'] = stats_circuits['ts_kind'].replace({'T': 'CL', 'NT': 'OL', 'na': 'base', '–': '–'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "281e1fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = stats_circuits.sort_values('circuit', key=lambda x: x.map({'–': 0, 'base': 1, 'OL': 2, 'CL': 3}))\n",
    "g = sns.catplot(data=plot_df, x='circuit', y='relative_expression', col='primers', hue='dox',\n",
    "                dodge=True, margin_titles=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4c806cf",
   "metadata": {},
   "source": [
    "While quick to make, the plots above are a bit hard to read/interpret. Let's make a couple nicer ones to see the final results of our analysis. Bonus points for statistical tests!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9314b2d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "circuit_palette = {'base': 'black', 'OL': 'gray', 'CL': 'teal', '–': 'black'}\n",
    "star_convert = list(reversed([(1.0, 'n.s.'), (0.05, '*'), (0.01, '**'), (0.001, '***'), (0.0001, '****')]))\n",
    "pval_to_stars = lambda x: next(val[1] for val in star_convert if val[0] > x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a559ddf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "f, axes = plt.subplots(1,3, figsize=(10,5), gridspec_kw={'wspace': 0.3}, sharex=True, sharey=True)\n",
    "primer_sets = ['mRuby2-1', 'mRuby2-2', 'FXN']\n",
    "for ax, primer in zip(axes, primer_sets):\n",
    "    plot_df = stats_circuits[stats_circuits.primers==primer]\n",
    "    y = 'relative_expression'\n",
    "    sns.scatterplot(data=plot_df, x='circuit', y=y, \n",
    "                    hue='circuit', palette=circuit_palette, ax=ax, legend=False)\n",
    "    ax.set(xlim=(-0.5,3.5), title=primer, ylabel='Relative expression, $2^{-\\Delta \\Delta C_p}$', ylim=(-0.05,1.3))\n",
    "    xval_order = [l.get_text() for l in ax.get_xticklabels()]\n",
    "\n",
    "    # T-tests comparing OL to CL circuits\n",
    "    pair = ('OL','CL')\n",
    "    x_start = xval_order.index(pair[0])\n",
    "    x_end = xval_order.index(pair[1])\n",
    "    y_loc = 0.9\n",
    "    pad = 0.02\n",
    "    star = pval_to_stars(sp.stats.ttest_ind(plot_df[(plot_df.circuit==pair[0])][y], plot_df[(plot_df.circuit==pair[1])][y],).pvalue)\n",
    "    print(f'{pair}: {star}')\n",
    "    _ = ax.text(x_start+(x_end-x_start)/2, y_loc - pad * (star.startswith('*')), star, \n",
    "                transform=ax.get_xaxis_transform(), ha='center', va='center', backgroundcolor='white', bbox={\"facecolor\": 'white', \"pad\": 1})\n",
    "    ax.add_artist(matplotlib.lines.Line2D([x_start, x_end], [y_loc, y_loc], transform=ax.get_xaxis_transform(), color='k', linewidth=1))\n",
    "    \n",
    "    fc = plot_df[(plot_df.circuit==pair[1])][y].mean() / plot_df[(plot_df.circuit==pair[0])][y].mean()\n",
    "    print(f'{fc:0.2f}')\n",
    "    _ = ax.text(x_start+(x_end-x_start)/2, y_loc+0.03, f'{fc:0.2f}x', size='small',\n",
    "                transform=ax.get_xaxis_transform(), ha='center', va='bottom', \n",
    "                backgroundcolor='white', bbox={\"facecolor\": 'white', \"pad\": 1})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b5beed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "f, axes = plt.subplots(1,4, figsize=(13,5), gridspec_kw={'wspace': 0.3}, sharex=True, sharey=True)\n",
    "primer_sets = ['EGFP-1', 'EGFP-2', 'FMRP-1', 'FMRP-2']\n",
    "for ax, primer in zip(axes, primer_sets):\n",
    "    plot_df = stats_circuits[stats_circuits.primers==primer]\n",
    "    y = 'relative_expression'\n",
    "    sns.scatterplot(data=plot_df, x='circuit', y=y, \n",
    "                    hue='circuit', palette=circuit_palette, ax=ax, legend=False)\n",
    "    ax.set(xlim=(-0.5,3.5), title=primer, ylabel='Relative expression, $2^{-\\Delta \\Delta C_p}$', ylim=(-0.05,3))\n",
    "    xval_order = [l.get_text() for l in ax.get_xticklabels()]\n",
    "\n",
    "    # T-tests comparing OL to CL circuits\n",
    "    pair = ('OL','CL')\n",
    "    x_start = xval_order.index(pair[0])\n",
    "    x_end = xval_order.index(pair[1])\n",
    "    y_loc = 0.9\n",
    "    pad = 0.02\n",
    "    star = pval_to_stars(sp.stats.ttest_ind(plot_df[(plot_df.circuit==pair[0])][y], plot_df[(plot_df.circuit==pair[1])][y],).pvalue)\n",
    "    print(f'{pair}: {star}')\n",
    "    _ = ax.text(x_start+(x_end-x_start)/2, y_loc - pad * (star.startswith('*')), star, \n",
    "                transform=ax.get_xaxis_transform(), ha='center', va='center', backgroundcolor='white', bbox={\"facecolor\": 'white', \"pad\": 1})\n",
    "    ax.add_artist(matplotlib.lines.Line2D([x_start, x_end], [y_loc, y_loc], transform=ax.get_xaxis_transform(), color='k', linewidth=1))\n",
    "    \n",
    "    fc = plot_df[(plot_df.circuit==pair[1])][y].mean() / plot_df[(plot_df.circuit==pair[0])][y].mean()\n",
    "    print(f'{fc:0.2f}')\n",
    "    _ = ax.text(x_start+(x_end-x_start)/2, y_loc+0.03, f'{fc:0.2f}x', size='small',\n",
    "                transform=ax.get_xaxis_transform(), ha='center', va='bottom', \n",
    "                backgroundcolor='white', bbox={\"facecolor\": 'white', \"pad\": 1})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62033205",
   "metadata": {},
   "source": [
    "# Example 2: Absolute quantification of a genomics library"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "574d94b6",
   "metadata": {},
   "source": [
    "For library quantification, we'll implement formulas from [NEB qPCR Library Quantification calculator](https://nebiocalculator.neb.com/#!/qPCRlibQnt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b77ccd96",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff74bcd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = rd.datadir/'data'/'qPCR'/'chris'/'2025.09.06-rcmc-lib-quant'\n",
    "output_path = rd.rootdir/'output'/'qPCR-example'\n",
    "cache_path = output_path/'data2.gzip'\n",
    "\n",
    "rd.plot.plot_well_metadata(base_path/'wells.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02edf7f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.DataFrame()\n",
    "if cache_path.exists(): data = pd.read_parquet(cache_path)\n",
    "else: \n",
    "    data = rd.qpcr.load_single_csv_with_metadata(base_path/'cp_table.txt', base_path/'wells.yaml', is_default=True)\n",
    "    data.dropna(subset='sample', inplace=True)\n",
    "    data['dilution'] = data['dilution'].astype(float)\n",
    "    data.to_parquet(rd.outfile(cache_path))\n",
    "display(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c1d1102",
   "metadata": {},
   "source": [
    "Plot samples (not standard curve) to look for outlier points."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c88a6b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = data[~(data['sample'].str.contains('standard')) & ~(data['sample']=='NTC')]\n",
    "f, ax = plt.subplots(1,1, figsize=(len(plot_df['sample'].unique())*0.4, 4))\n",
    "sns.stripplot(data=plot_df, x='sample', y='Cp', hue='dilution', ax=ax, dodge=True)\n",
    "sns.move_legend(ax, 'upper left', bbox_to_anchor=(1,1))\n",
    "_ = ax.set_xticklabels(ax.get_xticklabels(), rotation=90)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b7f212",
   "metadata": {},
   "source": [
    "## Compute standard curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd75042",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map standards to known concentrations (pM)\n",
    "standards_conc = {\n",
    "    'standard-1': 100,\n",
    "    'standard-2': 10,\n",
    "    'standard-3': 1,\n",
    "    'standard-4': 0.1,\n",
    "    'standard-5': 0.01,\n",
    "    'standard-6': 0.001,\n",
    "    'NTC': 0,\n",
    "}\n",
    "standards = data.loc[data.dilution==0, ['sample','Cp']]\n",
    "standards['conc'] = standards['sample'].map(standards_conc)\n",
    "standards['log_conc'] = standards.conc.apply(np.log10)\n",
    "\n",
    "# Perform linear regression on standards\n",
    "standard_curve = sp.stats.linregress(standards.loc[standards.conc!=0, 'log_conc'], standards.loc[standards.conc!=0, 'Cp'])\n",
    "print(f'y = {standard_curve.slope:.2f} x + {standard_curve.intercept:.2f}')\n",
    "print(f'R^2 = {abs(standard_curve.rvalue):.3f}')\n",
    "efficiency = (10**(-1/standard_curve.slope) - 1)*100 # percentage\n",
    "print(f'Efficiency: {efficiency:0.1f}% -- ' + ('good!' if (efficiency >= 90) and (efficiency <= 110) else 'outside optimal range'))\n",
    "\n",
    "# Plot standard curve\n",
    "f, ax = plt.subplots(1,1, figsize=(6,5))\n",
    "ax.scatter(x=standards.conc, y=standards.Cp, label='standards', ec='white', lw=0.75)\n",
    "xs = np.logspace(-3,2,1000)\n",
    "ys = standard_curve.slope * np.log10(xs) + standard_curve.intercept\n",
    "ax.plot(xs, ys, color='crimson', label='linear\\nregression')\n",
    "ax.set(xlabel='Conc. (pM)', xlim=(-0.0005,200), ylabel='Cp')\n",
    "ax.set_xscale('symlog', linthresh=0.001)\n",
    "ax.legend()\n",
    "pad = 0.02\n",
    "ax.annotate(f'$R^2$: {abs(standard_curve.rvalue):0.3f}', (0+pad, 0.1), xycoords='axes fraction',\n",
    "            ha='left', va='bottom', size='medium')\n",
    "ax.annotate(f'Efficiency: {efficiency:0.1f}%', (0+pad, 0+pad), xycoords='axes fraction',\n",
    "            ha='left', va='bottom', size='medium')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77bbc19e",
   "metadata": {},
   "source": [
    "Now, use the standard curve to quantify concentrations of samples (libraries).\n",
    "\n",
    "We'll take the mean of the technical replicates, since these looked good above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "535a36be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute sample concentrations from standard curve\n",
    "def calc_concentration(x, standard_curve):\n",
    "    return 10**((x.mean()-standard_curve.intercept)/standard_curve.slope)\n",
    "\n",
    "concentrations = data[~(data['sample'].str.contains('standard')) & ~(data['sample']=='NTC')].groupby(['sample','dilution'])['Cp'].apply(calc_concentration, standard_curve).reset_index().rename(columns={'Cp': 'conc'})\n",
    "display(concentrations)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d98dbc8c",
   "metadata": {},
   "source": [
    "## Calculate actual library concentration, adjusting for fragment size"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e0e718c",
   "metadata": {},
   "source": [
    "For this experiment, we ran the libraries on the Bioanalyzer. This gives us the average fragment size in the library, which we can use to convert concentration to mass. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63cd4c8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# From Bioanalyzer (manually type numbers (nM) from pdf of traces)\n",
    "frags = pd.DataFrame([\n",
    "    ['lib01', 425, 75.481], ['lib02', 426, 27.807], # UT -dox\n",
    "    ['lib03', 410, 34.314], ['lib04', 391, 48.973], # Con -dox\n",
    "    ['lib05', 397, 62.729], ['lib06', 433, 7.873],  # UT +dox\n",
    "    ['lib07', 434, 25.146], ['lib08', 427, 43.507], # Con +dox\n",
    "], columns=['sample', 'avg_frag_size', 'measured_conc'])\n",
    "\n",
    "frags['conc'] = frags['measured_conc'] * 10   # samples were diluted 1:10 for Bioanalyzer run\n",
    "display(frags)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1538fdf",
   "metadata": {},
   "source": [
    "Since the libraries look similar, let's just use the median across the samples as the average fragment size for all.\n",
    "\n",
    "Where does the 399 come from? I'm not sure XD We should look this up; maybe NEB has more info.\n",
    "\n",
    "Finally, we can convert the measured concentration to the actual library concentration using the dilution (and convert this to nM)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f667e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "median_frag_size = frags.avg_frag_size.median()\n",
    "print(f'Median avg. fragment size: {median_frag_size:n} bp')\n",
    "\n",
    "concentrations['avg_frag_size'] = median_frag_size\n",
    "concentrations['conc_size_adj'] = concentrations.conc * 399 / concentrations.avg_frag_size\n",
    "concentrations['lib_conc_size_adj'] = concentrations.conc_size_adj * concentrations.dilution / 1000 # convert from pM to nM\n",
    "concentrations['lib_conc'] = concentrations.conc * concentrations.dilution / 1000 # convert from pM to nM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9043008",
   "metadata": {},
   "source": [
    "We can plot this value for each dilution. These should agree if we pipetted well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d02521cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = concentrations\n",
    "f, ax = plt.subplots(1,1, figsize=(len(plot_df['sample'].unique())*0.4, 4))\n",
    "sns.stripplot(data=plot_df, x='sample', y='lib_conc_size_adj', hue='dilution', ax=ax, dodge=True)\n",
    "sns.move_legend(ax, 'upper left', bbox_to_anchor=(1,1))\n",
    "_ = ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "_ = ax.set(ylabel='Size-adjusted conc. (nM)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f8d968",
   "metadata": {},
   "source": [
    "Average measurements over the two dilutions to get a single library concentration for each sample (library)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca341458",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_library_conc = concentrations.groupby('sample')[['lib_conc','lib_conc_size_adj']].mean().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "500989c1",
   "metadata": {},
   "source": [
    "Note that the Bioanalyzer run also calculated a concentration. This is size-adjusted in nM, which is the same as what we calculated for our final library concentration. Let's plot to compare; we expect these to generally agree, but they may not be perfect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b35552d",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = sns.scatterplot(data=final_library_conc, x='sample', y='lib_conc_size_adj', label='qPCR')\n",
    "sns.scatterplot(data=frags, x='sample', y='conc', ax=f, label='Bioanalyzer')\n",
    "_ = f.set_xticklabels(f.get_xticklabels(), rotation=90)\n",
    "f.set(ylabel='Library conc. (nM)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4fc25df",
   "metadata": {},
   "source": [
    "Looks like the Bioanalyzer concentrations are pretty variable and different from the qPCR concentrations. Some of the samples had incorrect upper marker detection (i.e., size range in which to quantify) in the Bioanalyzer results, which we forgot to fix, so trust the qPCR values more."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82c017fc",
   "metadata": {},
   "source": [
    "## Compute masses for pooling RCMC libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0efe0247",
   "metadata": {},
   "source": [
    "For RCMC, we prepare libraries for each sample, then pool the samples for the region-capture step. The pooling should be equal by mass for each library.\n",
    "\n",
    "We'll use another formula from NEB to convert moles to mass. This requires the measured concentrations (not the size-adjusted one, as the formula does this) and the known volume of the libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b988198",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_moles_to_mass(moles, length):\n",
    "    \"\"\"\n",
    "    For a given amount of DNA in moles, use its length\n",
    "    to calculate its mass in grams.\n",
    "\n",
    "    Formula from NEB: \n",
    "    g = mol x (bp x 615.94 g/mol/bp + 36.04 g/mol)\n",
    "     - mass of dsDNA (g) = moles dsDNA x (molecular weight of dsDNA (g/mol))\n",
    "     - molecular weight of dsDNA = (number of base pairs of dsDNA x average molecular weight of a base pair) + 36.04 g/mol\n",
    "     - average molecular weight of a base pair = 615.94 g/mol, excluding the water molecule removed during polymerization \n",
    "       and assuming deprotonated phosphate hydroxyls\n",
    "     - the additional 36.04 g/mol accounts for the 2 -OH and 2 -H added back to the ends\n",
    "     - bases are assumed to be unmodified\n",
    "    \"\"\"\n",
    "    return moles * (length * 615.96 + 36.04)\n",
    "\n",
    "lib_volume = 25 / 1e6   # convert uL to L\n",
    "final_library_conc['moles'] = lib_volume * final_library_conc['lib_conc'] / 1e9 # convert nM to moles\n",
    "final_library_conc['mass_ug'] = final_library_conc['moles'].apply(convert_moles_to_mass, [median_frag_size]) * 1e6\n",
    "display(final_library_conc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5261bc6a",
   "metadata": {},
   "source": [
    "Lastly, compute the volume to pipet to obtain 500 ng (0.5 µg) per library. This is how much we want in the pool."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b86a8f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_library_conc['500ng_vol_uL'] = 0.5 / final_library_conc['mass_ug'] * lib_volume * 1e6\n",
    "print(f'Pooled library volume: {final_library_conc['500ng_vol_uL'].sum():0.1f} µL')\n",
    "\n",
    "# Print summary for pipetting\n",
    "display(final_library_conc[['sample','500ng_vol_uL']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d616e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
